src/GNDStk/Node/src/subscript.hpp

```diff
@@ -0,0 +1,275 @@
+wish to produce lvalues, which allow you to write to - not just read from - a
+(non-const) GNDStk tree or node. You can use operator[] as a "right-hand side"
+too, just as you can with operator(), but should do so with the understanding
+that it won't play well with "auto," or with similar auto-like processes such
+as a compiler's determination of a function template parameter's type.
+*/
+
+
+
+// -----------------------------------------------------------------------------
+// node[meta_t]
+// -----------------------------------------------------------------------------
+
+// The <void> cases are special and can just fall back on operator(), which
+// gives [const] std::string *reference* (thus lvalue if non-const) already.
+
+
+// ------------------------
+// <void>
+// ------------------------
+
+// <void> const
+const std::string &operator[](const meta_t<void> &kwd) const
+{
+   return (*this)(kwd);
+}
+
+// <void>
+std::string &operator[](const meta_t<void> &kwd)
+{
+   return (*this)(kwd);
+}
+
+
+// ------------------------
+// <TYPE>
+// ------------------------
+
+// <TYPE> const
+template<class TYPE, class CONVERTER>
+auto operator[](const meta_t<TYPE,CONVERTER> &kwd) const
+{
+   // meta_ref<CONST=true>, in contrast to meta_ref<CONST=false>, dispenses
+   // with the assignment operators of the latter class, for obvious reasons.
+   // You can still use the result of this function as an rvalue, though, and
+   // it has the same conversion operators that meta_ref<CONST=false> has.
+   return detail::meta_ref<Node,true, TYPE,CONVERTER>(kwd,*this);
+}
+
+// <TYPE>
+template<class TYPE, class CONVERTER>
+auto operator[](const meta_t<TYPE,CONVERTER> &kwd)
+{
+   return detail::meta_ref<Node,false,TYPE,CONVERTER>(kwd,*this);
+}
+
+
+
+// -----------------------------------------------------------------------------
+// node[child_t]
+// -----------------------------------------------------------------------------
+
+/*
+For now, we won't make or use our lvalue-enabling class, child_ref, for the
+case of allow::many, because it isn't clear what semantics we'd want when the
+lvalue represents a container of some kind. Say, for example, that we're in
+an <axes> node n, and wrote:
+
+   n[axis];
+
+where axis (not axes) is of course an allow::many. The call operator analog,
+
+   n(axis)
+
+returns a container (default std::vector) of nodes. We could have child_ref in
+this case contain, for example, a std::vector<std::reference_wrapper<node>>,
+or even just a std::vector<node*>. As an lvalue, though, what would assignment
+mean? Replace each container element, requiring that the assignment's RHS has
+the same container size? Remove all the existing axis nodes of the parent node,
+completely replacing them with the nodes in the assignment's RHS? Until/unless
+we decide what's best, I'll have operator[] for allow::many child_t objects do
+what operator() does. So, fow now: defined, but not usable as an lvalue.
```

Comment:
Now, if I wrote the following:
```
n[ axis, [] ( const auto& node ) { return node( index ) == "0"; } ]
```
would that return an lvalue so that I can actually modify that node?

Reply:
I'm inclined to agree with you that if the operator[] returns more than one node, we should not treat the container as an lvalue. Having the vector contain reference_wrappers is a more sensible solution - unless we establish the rule that ONLY a single node can be an lvalue.

Reply:
I'm thinking more and more than the call operator should also return a vector of reference wrappers, not a vector of copies of the nodes.

And it goes without saying (although, well, I'm saying it) that that applies to the case where you use child_t<void>, not child_t<some type other than void>. In the latter case, conversion from node to some user type must be provided, and the conversion is applied to create each element of the output vector. So stuff is created on-the-fly, and the reference idea is not applicable.

Reply:
> Now, if I wrote the following:
> 
> ```
> n[ axis, [] ( const auto& node ) { return node( index ) == "0"; } ]
> ```
> 
> would that return an lvalue so that I can actually modify that node?

I think you want to *downgrade* axis to an allow::one, and *add* the lambda:
```
n[ --axis + lambda ];
```
Let me know if this indeed works. While I think I got `operator[]` right, I don't have full tests for it yet.

Reply:
Of course you'd presumably actually assign that to something:
```
n[ --axis + lambda ] = foobar;
```
Note that my `operator[]` is tricked out so that it can assign to either a value of the *type* given to (in this case) `axis`, or to anything that's implicitly convertible to that type (like, `int` to `double`) - or even to a string, if for whatever reason you'd like to put in the raw string directly instead of using the type you've associated with the keyword.

Reply:
Ah, I forgot that I needed a '+' to add the filter. I'll give it a try when I update the v1.9 branch.

Reply:
Re: my above comment edit. If I ever make a comment that sounds stupid, let me know, and I can probably fix it. It's probably markdown's fault!!. Unless it's my fault, in which case I'll still try to fix it.

Reply:
So if I write:
```
auto filter = [] ( const auto& node ) { return node( index ) == "0"; };
n[ Axis{} / --axis + filter ] = Axis( 0, "someLabel", "someUnit" );
```
I can use one of the higher level components to change the node itself (if I have a convert function handy, that is)?

Reply:
By the way, regarding my `--` and `++` operators on keywords: Strangely enough, they *don't* actually modify the object itself, like how `--i` decrements to actual contents of the integer `i`. My keyword arithmetic always produces brand-new objects. (In many cases it couldn't *not* do that, because the arithmetic changes a template type!). So, `--child_t` has no effect on the original, and the original can even be `const`. Note that the language does indeed allow this behavior - of increment and decrement not changing the object to which they're applied. Of course I wouldn't do it if the language didn't allow it.

Reply:
I need to check if the remarks in my `foobar` example above actually work as I indent, re: the right-hand side being `std::string`. I may need to explicitly support the `char*` case, to prevent a general template function overload (from which an exact match can be produced) from being called, in place of the expected implicit `char *` to `std::string` conversion. Ah, technical stuff.

Reply:
To prevent any possible confusion... Be aware that a single minus (unary minus, `-`) in front of a keyword downgrades the *first* template argument - `type` - to `void`, meaning raw tree form rather than user-defined type. That applies to both `meta_t` and `child_t`. *Two* minus signs (predecrement, `--`) in front of a keyword downgrades the *second* template argument of a `child_t` - making it `allow::one`, regardless of its previous value. (The predecrement doesn't apply to `meta_t` objects, which don't have the `one` versus `many` distinction.)

There's indeed a pattern to all this funny stuff I wrote. A user just needs to stare at it for a bit, in order for it all to make sense. If it still doesn't make sense after some study, then let's talk about it. :-)

Reply:
> So if I write:
> 
> ```
> auto filter = [] ( const auto& node ) { return node( index ) == "0"; };
> n[ Axis{} / --axis + filter ] = Axis( 0, "someLabel", "someUnit" );
> ```
> 
> I can use one of the higher level components to change the node itself (if I have a convert function handy, that is)?

Offhand that looks correct, and it *should* work. Let me know if it doesn't! The operator business is fairly recent, and the bracket operator is brand-new, so there could always be a glitch here or there.

Of course we can, as I illustrated in my sample high-level Axis, associate an existing node with a high-level structure, and have a `save()` or other operation that changes the node automatically. We'll probably want to allow more than one way to do things.

Your comment reminds me: we should have a discussion sometime of what I did provide in the way of built-in `convert` functions. Roughly speaking: a handful of basic ones are provided for metadata (string/type conversions), and none are provided for children (node/type) conversions. None, because I can't imagine any meaningful default for that case; it depends on what's in somebody's data structure.

Reply:
It looks like all this operator business has proven to be even more useful than I anticipated it might be when I wrote it. That's a good thing.